File: mytradebot/trading/exchange_models.py
"""文件用途：交易所统一数据模型，定义持仓、订单和下单结果。"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True)
class ExchangePosition:
    """说明：交易所实时持仓。"""

    exchange: str
    symbol: str
    instrument_key: str
    side: str
    size: float
    entry_price: float
    mark_price: float
    unrealized_pnl: float
    leverage: float | None = None
    margin: float | None = None
    liquidation_price: float | None = None

    def to_payload(self) -> dict[str, Any]:
        return {
            "exchange": self.exchange,
            "symbol": self.symbol,
            "instrumentKey": self.instrument_key,
            "side": self.side,
            "size": self.size,
            "entryPrice": self.entry_price,
            "markPrice": self.mark_price,
            "unrealizedPnl": self.unrealized_pnl,
            "leverage": self.leverage,
            "margin": self.margin,
            "liquidationPrice": self.liquidation_price,
        }


@dataclass(frozen=True)
class ExchangeOrder:
    """说明：交易所挂单/活跃订单。"""

    exchange: str
    symbol: str
    instrument_key: str
    order_id: str
    side: str
    order_type: str
    size: float
    price: float | None
    filled_size: float
    status: str
    created_at_ms: int

    def to_payload(self) -> dict[str, Any]:
        return {
            "exchange": self.exchange,
            "symbol": self.symbol,
            "instrumentKey": self.instrument_key,
            "orderId": self.order_id,
            "side": self.side,
            "orderType": self.order_type,
            "size": self.size,
            "price": self.price,
            "filledSize": self.filled_size,
            "status": self.status,
            "createdAtMs": self.created_at_ms,
        }


@dataclass(frozen=True)
class OrderResult:
    """说明：下单结果。"""

    exchange: str
    order_id: str | None = None
    average_price: float | None = None
    filled_size: float | None = None
    resting: bool = False
    error: str | None = None
    raw: dict[str, Any] = field(default_factory=dict)

    @property
    def ok(self) -> bool:
        return self.error is None

    def to_payload(self) -> dict[str, Any]:
        return {
            "exchange": self.exchange,
            "orderId": self.order_id,
            "averagePrice": self.average_price,
            "filledSize": self.filled_size,
            "resting": self.resting,
            "error": self.error,
        }

File: tradex/trading/exchange-models.ts
export interface ExchangePosition {
  exchange: string;
  symbol: string;
  instrumentKey: string;
  side: string;
  size: number;
  entryPrice: number;
  markPrice: number;
  unrealizedPnl: number;
  leverage?: number | null;
  margin?: number | null;
  liquidationPrice?: number | null;
}

export interface ExchangeOrder {
  exchange: string;
  symbol: string;
  instrumentKey: string;
  orderId: string;
  side: string;
  orderType: string;
  size: number;
  price: number | null;
  filledSize: number;
  status: string;
  createdAtMs: number;
  reduceOnly?: boolean | null;
  triggerPrice?: number | null;
  tpsl?: string | null;
}

export interface TradeSyncResult {
  exchange: string;
  status: string;
  closed: boolean;
  reason: string;
  position: ExchangePosition | null;
  activeOrders: ExchangeOrder[];
  error: string | null;
}

export interface OrderResult {
  exchange: string;
  orderId: string | null;
  averagePrice: number | null;
  filledSize: number | null;
  resting: boolean;
  error: string | null;
  raw: Record<string, unknown>;
}

export function orderOk(result: OrderResult): boolean {
  return result.error === null;
}

export function orderResult(input: Partial<OrderResult> & { exchange: string }): OrderResult {
  return {
    exchange: input.exchange,
    orderId: input.orderId ?? null,
    averagePrice: input.averagePrice ?? null,
    filledSize: input.filledSize ?? null,
    resting: input.resting ?? false,
    error: input.error ?? null,
    raw: input.raw ?? {},
  };
}

export function positionToPayload(position: ExchangePosition): Record<string, unknown> {
  return {
    exchange: position.exchange,
    symbol: position.symbol,
    instrumentKey: position.instrumentKey,
    side: position.side,
    size: position.size,
    entryPrice: position.entryPrice,
    markPrice: position.markPrice,
    unrealizedPnl: position.unrealizedPnl,
    leverage: position.leverage ?? null,
    margin: position.margin ?? null,
    liquidationPrice: position.liquidationPrice ?? null,
  };
}

export function orderToPayload(order: ExchangeOrder): Record<string, unknown> {
  return {
    exchange: order.exchange,
    symbol: order.symbol,
    instrumentKey: order.instrumentKey,
    orderId: order.orderId,
    side: order.side,
    orderType: order.orderType,
    size: order.size,
    price: order.price,
    filledSize: order.filledSize,
    status: order.status,
    createdAtMs: order.createdAtMs,
    ...(order.reduceOnly !== undefined ? { reduceOnly: order.reduceOnly } : {}),
    ...(order.triggerPrice !== undefined ? { triggerPrice: order.triggerPrice } : {}),
    ...(order.tpsl !== undefined ? { tpsl: order.tpsl } : {}),
  };
}

export function syncResultToPayload(result: TradeSyncResult): Record<string, unknown> {
  return {
    exchange: result.exchange,
    status: result.status,
    closed: result.closed,
    reason: result.reason,
    position: result.position ? positionToPayload(result.position) : null,
    activeOrders: result.activeOrders.map(orderToPayload),
    error: result.error,
  };
}
